File: 2215-find-the-difference-of-two-arrays/2215-find-the-difference-of-two-arrays.go
type stats struct {
    first bool
    second bool
}

func (s *stats) update(i int) {
    if i == 1 {
        s.first = true
    } else {
        s.second = true
    }
}

func findDifference(nums1 []int, nums2 []int) [][]int {
    
    hashMap := make(map[int]*stats)
    for i := range nums1 {
        if _, ok := hashMap[nums1[i]]; !ok {
            hashMap[nums1[i]] = &stats{}
        }
        hashMap[nums1[i]].update(1)
    }
    for i := range nums2 {
        if _, ok := hashMap[nums2[i]]; !ok {
            hashMap[nums2[i]] = &stats{}
        }
        hashMap[nums2[i]].update(2)
    }
    var ans = make([][]int,2)
    ans[0] = make([]int,0)
    ans[1] = make([]int,0)
    for k, v := range hashMap {
        if !(v.first && v.second) {
            if !v.first {
                ans[1] = append(ans[1], k)
            }
            if !v.second {
                ans[0] = append(ans[0], k)
            }
        }
    }
    return ans
}

File: 2215-find-the-difference-of-two-arrays/2215-find-the-difference-of-two-arrays.java
import java.util.List;
import java.util.ArrayList;

class Solution {
    public List<List<Integer>> findDifference(int[] nums1, int[] nums2) {
        HashMap<Integer,Boolean> hashMap1 = arrayToHashMap(nums1);
        HashMap<Integer,Boolean> hashMap2 = arrayToHashMap(nums2);
        HashMap<Integer,Boolean> intersection = hashMapIntersection(hashMap1, hashMap2);

        List<List<Integer>> answer = new ArrayList<>();
        int[][] nums = {nums1, nums2};
        for (int[] num : nums) {
            HashMap<Integer,Boolean> innerHashMap = new HashMap<>();
            for (int n : num) {
                if (!intersection.containsKey(n)) {
                    innerHashMap.put(n, true);
                }
            }
            answer.add(new ArrayList<>(innerHashMap.keySet()));
        }
        return answer;
    }

    static public HashMap<Integer,Boolean> arrayToHashMap(int[] nums) {
        HashMap<Integer,Boolean> hashMap = new HashMap<>();
        for (int num : nums) {
            hashMap.put(num, true);
        }
        return hashMap;
    }

    static public HashMap<Integer,Boolean> hashMapIntersection(HashMap<Integer,Boolean> hashMap1, HashMap<Integer,Boolean> hashMap2) {
        HashMap<Integer,Boolean> hashMap = new HashMap<>();
        for (Map.Entry<Integer, Boolean> entry : hashMap1.entrySet()) {
            Integer key = entry.getKey();
            if (hashMap2.containsKey(key)) {
                hashMap.put(key, true);
            }
        }
        return hashMap;
    }
}